### PR TITLE
Street address length limit

### PR DESCRIPTION
**Pull request by mccottry**
https://app.asana.com/0/search/1107548184740967/1001491688938165

This is designed to fix the problem of extra long street address lines that some merchants are experiencing.  If any address street address line is greater than 50 characters, we attempt to shift if possible, if not, concatenate, and if all else does not resolve the problem, clip so that the address can be processed.

### CODE REVIEW

File: app/code/community/Bolt/Boltpay/Model/BoltOrder.php [hunk 1]
```diff
@@ -365,16 +365,135 @@ protected function getCorrectedTotal($projectedTotal, $magentoDerivedCartData)
...
+
+                return;
+            }
+
+            ///////////////////////////////////////////////////////////
+            /// Try to find a natural splitting point in address line,
+            /// then split the string
+            ///////////////////////////////////////////////////////////
+            $breakPoint = strpos($tempLine, ' ', strlen($tempLine) - 50);
+
```
Comment (danac-gs) [inline app/code/community/Bolt/Boltpay/Model/BoltOrder.php]: Not sure if the logic is good here. I tried it with: 
```php
$tempLine = "1234567890 1234567890 1234567890 1234567890 123456 7890 1234567890";
```
Try this:
```php
$offset = -(strlen($tempLine) - 50);
$breakPoint = strrpos ($tempLine , ' ', $offset); 
```

Comment (mccottry) [reply app/code/community/Bolt/Boltpay/Model/BoltOrder.php]: @danac-gs  I think you reviewed code that was a work in progress.   The WIP will come off in a moment.  Please evaluate that.

**Comment by daisy1754** [comment]
Do we know if this validation is built into magento and all merchants have 50 character limit? If different merchants may have different length limit, I'd rather want to explore generic way of allowing merchant to override address

File: app/code/community/Bolt/Boltpay/Model/BoltOrder.php [hunk 2]
```diff
@@ -365,17 +365,132 @@ protected function getCorrectedTotal($projectedTotal, $magentoDerivedCartData)
...
+            if ($this->hasEmptyAddressLine($streetAddressArray, $startingIndex)) {
+
+                // Since there is an empty space, we fill it by shifting
+
+                for ($i = $startingIndex + 1; $i < count($streetAddressArray); $i++) {
+                    $nextAddressLine = $streetAddressArray[$i];
+                    if ($extraAddressInfo) $streetAddressArray[$i] = $extraAddressInfo;
+
+                    $extraAddressInfo = preg_replace('/\n/', '', $nextAddressLine);
```
Comment (matt-thomason) [inline app/code/community/Bolt/Boltpay/Model/BoltOrder.php]: Should this `preg_replace` be done before the potential assignment?

File: app/code/community/Bolt/Boltpay/Model/BoltOrder.php [hunk 3]
```diff
@@ -365,17 +365,132 @@ protected function getCorrectedTotal($projectedTotal, $magentoDerivedCartData)
...
+            $extraAddressInfo = substr($tempLine, $breakPoint + 1);
+            ///////////////////////////////////////////////////////////
+
+            $streetAddressArray[$startingIndex] = $correctedAddressLine;
+
+            if ($this->hasEmptyAddressLine($streetAddressArray, $startingIndex)) {
+
+                // Since there is an empty space, we fill it by shifting
+
+                for ($i = $startingIndex + 1; $i < count($streetAddressArray); $i++) {
```
Comment (matt-thomason) [inline app/code/community/Bolt/Boltpay/Model/BoltOrder.php]: We're going to get an OOB error here I think if we start at higher indices. Can we do a `foreach` on `array_slice($streetAddressArray, $startingIndex + 1)` instead?

File: tests/unit/testsuite/Bolt/Boltpay/Model/BoltOrderTest.php [hunk 4]
```diff
@@ -329,4 +329,156 @@ public function testCorrectBillingAddressWithNoBillingAddress() {
...
+        $this->assertLessThan(51, strlen($streetAddressData[0]));
+        $this->assertNotEmpty($streetAddressData[1]);
+        $this->assertLessThan(51, strlen($streetAddressData[1]));
+        $this->assertNotEmpty($streetAddressData[2]);
+        $this->assertLessThan(51, strlen($streetAddressData[2]));
+        $this->assertNotEmpty($streetAddressData[3]);
+        $this->assertLessThan(51, strlen($streetAddressData[3]));
+        $this->assertEquals('Special Attention: Velvet Fox Group, Thurston Hall', $streetAddressData[3]);
+
+    }
```
Comment (matt-thomason) [inline tests/unit/testsuite/Bolt/Boltpay/Model/BoltOrderTest.php]: Good tests 👍

File: app/code/community/Bolt/Boltpay/Model/BoltOrder.php [hunk 5]
```diff
@@ -365,17 +365,132 @@ protected function getCorrectedTotal($projectedTotal, $magentoDerivedCartData)
...
+
+                return;
+            }
+
+            ///////////////////////////////////////////////////////////
+            /// Try to find a natural splitting point in address line,
+            /// then split the string
+            ///////////////////////////////////////////////////////////
+            $breakPoint = strrpos(substr($tempLine, 0, 50), ' ');
+
```
Comment (danac-gs) [inline app/code/community/Bolt/Boltpay/Model/BoltOrder.php]: To cover the edge case where 51st character is space (exactly 50 remain for the line) you should change 50 to 51.

**Comment by mccottry** [comment]
This will be handled by custom filters instead.  This branch will be left open temporarily as it can be used as an implementation for this case for effected merchants.

**Pull request closed without merging**